### PR TITLE
{Network} Add new features for `custom-ip prefix`, `public-ip` and `public-ip prefix` 

### PR DESCRIPTION
**Pull request by 00Kai0**
Closed https://github.com/Azure/azure-cli/issues/18953

--

**Description**

* add a new family of commands `az network custom-ip prefix`
* update new parameters for `az network public-ip` and `az network public-ip prefix`

**Testing Guide**

**History Notes**

[Network]  Add a new command group `az network custom-ip prefix`.
[Network] `az network public-ip`: Add parameter `--ip-address`.
[Network] `network public-ip prefix create`: Add parameter `--custom-ip-prefix-name`.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 1]
```diff
@@ -4701,6 +4701,61 @@
     crafted: true
 """
 
+helps['network custom-ip'] = """
+type: group
+short-summary: Manage custom IP prefix
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:4706]:
```suggestion
short-summary: Manage custom IP
```

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 2]
```diff
@@ -4701,6 +4701,61 @@
...
+helps['network custom-ip prefix'] = """
+type: group
+short-summary: Manage custom IP prefix resources.
+"""
+
+helps['network custom-ip prefix create'] = """
+type: command
+short-summary: Create a custom IP prefix resource.
+examples:
+  - name: Create a custom IP prefix resource. (autogenerated)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:4718]:
```suggestion
  - name: Create a custom IP prefix resource.
```

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 3]
```diff
@@ -4701,6 +4701,61 @@
...
+    text: |
+        az network custom-ip prefix create --location westus2 --name MyCustomIpPrefix --resource-group MyResourceGroup
+    crafted: true
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:4721]:
```suggestion
```

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 4]
```diff
@@ -4701,6 +4701,61 @@
...
+
+helps['network custom-ip prefix update'] = """
+type: command
+short-summary: Update a custom IP prefix resource.
+examples:
+  - name: Update a custom IP prefix resource. (autogenerated)
+    text: |
+        az network custom-ip prefix update --name MyCustomIpPrefix --resource-group MyResourceGroup --set useRemoteGateways=true
+    crafted: true
+"""
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:4757]:
```suggestion
helps['network custom-ip prefix delete'] = """
type: command
short-summary: Delete a custom IP prefix resource.
examples:
  - name: Delete a custom IP prefix resource.
    text: |
        az network custom-ip prefix delete --name MyCustomIpPrefix --resource-group MyResourceGroup
"""

helps['network custom-ip prefix list'] = """
type: command
short-summary: List custom IP prefix resources.
"""

helps['network custom-ip prefix show'] = """
type: command
short-summary: Get the details of a custom IP prefix resource.
examples:
  - name: Get the details of a custom IP prefix resource.
    text: |
        az network custom-ip prefix show --name MyCustomIpPrefix --resource-group MyResourceGroup --subscription MySubscription
"""

helps['network custom-ip prefix update'] = """
type: command
short-summary: Update a custom IP prefix resource.
examples:
  - name: Update a custom IP prefix resource.
    text: |
        az network custom-ip prefix update --name MyCustomIpPrefix --resource-group MyResourceGroup --set useRemoteGateways=true
"""
```

File: src/azure-cli/azure/cli/command_modules/network/_params.py [hunk 5]
```diff
@@ -1724,6 +1736,7 @@ def load_arguments(self, _):
         c.argument('zone', zone_type, min_api='2017-06-01', max_api='2020-07-01')
         c.argument('zone', zone_compatible_type, min_api='2020-08-01')
         c.argument('ip_tags', nargs='+', min_api='2017-11-01', help="Space-separated list of IP tags in 'TYPE=VAL' format.", validator=validate_ip_tags)
+        c.argument('ip_address', help='The IP address associated with the public IP address resource.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_params.py:1739]: please add min_api for it

File: src/azure-cli/azure/cli/command_modules/network/commands.py [hunk 6]
```diff
@@ -444,6 +444,12 @@ def load_command_table(self, _):
         min_api='2020-05-01'
     )
 
+    network_custom_ip_prefix_sdk = CliCommandType(
+        operations_tmpl='azure.mgmt.network.operations#CustomIPPrefixesOperations.{}',
+        client_factory=cf_custom_ip_prefixes,
+        min_api='2021-02-01'
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/commands.py:450]: Its min_api does't match with `2020-06-01` used in comman_group definition.

File: src/azure-cli/azure/cli/command_modules/network/_params.py (hunk 5 above)
Comment (00Kai0) [reply src/azure-cli/azure/cli/command_modules/network/_params.py:1739]: It's always there from the  beginning: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/network/azure-mgmt-network/azure/mgmt/network/v2015_06_15/models/_models_py3.py#L3083

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 7]
```diff
@@ -5809,11 +5813,42 @@ def run_network_configuration_diagnostic(cmd, client, watcher_rg, watcher_name,
...
+        tags=tags,
+        signed_message=signed_message,
+        authorization_message=authorization_message
+    )
+
+    if custom_ip_prefix_parent:
+        try:
+            prefix.custom_ip_prefix_parent = client.get(resource_group_name, custom_ip_prefix_name)
+        except ResourceNotFoundError:
+            raise ResourceNotFoundError("Custom ip prefix parent {} is'nt exist".format(custom_ip_prefix_name))
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:5835]:
```suggestion
            raise ResourceNotFoundError("Custom ip prefix parent {} isn't exist".format(custom_ip_prefix_name))
```

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 8]
```diff
@@ -5809,11 +5813,42 @@ def run_network_configuration_diagnostic(cmd, client, watcher_rg, watcher_name,
...
+
+    return client.begin_create_or_update(resource_group_name, custom_ip_prefix_name, prefix)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:5837]: It is a long running operation. Please support `--no-wait` and wrap this line by `sdk_no_wait` function

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 9]
```diff
@@ -5809,11 +5813,42 @@ def run_network_configuration_diagnostic(cmd, client, watcher_rg, watcher_name,
...
+
+
+def update_custom_ip_prefix(instance, tags=None):
+    if tags is not None:
+        instance.tags = tags
+    return instance
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:5843]: Why this function only support tags? Can user update other parameters used in create command?

Comment (00Kai0) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:5843]: just follow the powershell: https://docs.microsoft.com/en-us/powershell/module/az.network/update-azcustomipprefix?view=azps-6.2.1

Comment (kairu-ms) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:5843]: Please add other arguments which can be updated as well.

File: src/azure-cli/azure/cli/command_modules/network/commands.py [hunk 10]
```diff
@@ -1207,6 +1213,15 @@ def _make_singular(value):
         g.custom_show_command('show', 'show_nw_troubleshooting_result', validator=process_nw_troubleshooting_show_namespace)
     # endregion
 
+    # region CustomIpPrefix
+    with self.command_group('network custom-ip prefix', network_custom_ip_prefix_sdk, client_factory=cf_custom_ip_prefixes, is_preview=True, min_api='2020-06-01') as g:
+        g.custom_command('create', 'create_custom_ip_prefix', supports_no_wait=True)
+        g.command('delete', 'begin_delete')
+        g.custom_command('list', 'list_custom_ip_prefixes')
+        g.show_command('show')
+        g.generic_update_command('update', setter_name='begin_create_or_update', custom_func_name='update_custom_ip_prefix')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/commands.py:1222]:
```suggestion
        g.generic_update_command('update', setter_name='begin_create_or_update', custom_func_name='update_custom_ip_prefix', supports_no_wait=True)
```

**Review by kairu-ms — APPROVED** [review]

**Review by kairu-ms — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**